### PR TITLE
AEROGEAR-2453 Implement storage and credential managers for iOS

### PR DESCRIPTION
**Pull request by aidenkeating**
## Motivation

JIRA: https://issues.jboss.org/browse/AEROGEAR-2453

## Description

Implement the storage and credential managers for iOS.

## Progress

- [x] Implementation
- [x] Unit tests for managers
- [x] Move iOS unit tests
- [x] Documentation

### CODE REVIEW

File: Auth/Auth.Platform.iOS/Credentials/OIDCCredential.cs [hunk 1]
```diff
@@ -0,0 +1,49 @@
+﻿using System;
+using Foundation;
+using OpenId.AppAuth;
+
+namespace AeroGear.Mobile.Auth.Credentials
```
Comment (aidenkeating) [inline Auth/Auth.Platform.iOS/Credentials/OIDCCredential.cs]: This needs documentation

File: Core/Core.Platform.iOS/Storage/KeychainWrapper.cs [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+        /// <summary>
+        /// Save the specified value using the specified key as an identifier.
+        /// This key can later be used to retrieve and remove the value.
+        /// </summary>
+        /// <returns><c>true</c> if the save operation succeeded. Otherwise <c>false</c>.</returns>
+        /// <param name="key">Identifier.</param>
+        /// <param name="value">Value to store.</param>
+        public bool Save(string key, string value)
+        {
+            Console.WriteLine("=== SAVE ===");
```
Comment (aidenkeating) [inline Core/Core.Platform.iOS/Storage/KeychainWrapper.cs]: Remove

File: Core/Core.Platform.iOS/Storage/KeychainWrapper.cs [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+            bool val = SetPassword(KeychainService, key, value, SecAccessible.Always, DefaultSynchronize) == SecStatusCode.Success;
+            Console.WriteLine(val);
```
Comment (aidenkeating) [inline Core/Core.Platform.iOS/Storage/KeychainWrapper.cs]: Remove and just return the response

File: Core/Core.Platform.iOS/Storage/KeychainWrapper.cs [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+            SecRecord secRecord = new SecRecord(SecKind.GenericPassword)
+            {
+                Service = service,
+                Account = username,
+                Synchronizable = synchronizable
+            };
+            SecStatusCode secStatusCode;
+            SecRecord queryRecord = SecKeyChain.QueryAsRecord(secRecord, out secStatusCode);
+
+            Console.WriteLine("=== SUPER DUPER ===");
```
Comment (aidenkeating) [inline Core/Core.Platform.iOS/Storage/KeychainWrapper.cs]: Remove

File: Core/Core.Platform.iOS/Storage/KeychainWrapper.cs [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+        private string GetPassword(string service, string username, bool synchronizable)
+        {
+            if (service == null)
+            {
+                throw new ArgumentNullException("service", "Service cannot be null");
+            }
+            if (username == null)
+            {
+                throw new ArgumentNullException("username", "Username cannot be null");
+            }
```
Comment (wei-lee) [inline Core/Core.Platform.iOS/Storage/KeychainWrapper.cs]: can we use the methods in the SanityCheck class instead?

Comment (aidenkeating) [reply Core/Core.Platform.iOS/Storage/KeychainWrapper.cs]: Ah yep, forgot about those

**Comment by aidenkeating** [comment]
@wei-lee I can't find any hard limit from a quick search. Looks to be limited more by `NSString` which is very large. I'll do a more thorough search though to make sure.

**Comment by wei-lee** [comment]
@aidenkeating it's ok. I think it's good for our use case. But based on the actual implementation, should we rename the class to `SecureKeyValueStore` to make it more clear? Just a thought...

**Comment by aidenkeating** [comment]
@wojta Mind taking a look?

**Comment by aidenkeating** [comment]
@ziccardi Mind taking a look?

**Comment by aidenkeating** [comment]
@wei-lee Mind taking a look?

File: Auth/Auth.Platform.iOS/Credentials/OIDCCredential.cs (hunk 1 above)
Comment (wei-lee) [reply Auth/Auth.Platform.iOS/Credentials/OIDCCredential.cs]: is this a duplicated impl in #23 ?

File: Example/Example.Android/Example.Android.csproj [hunk 6]
```diff
@@ -67,6 +67,36 @@
...
+    </Reference>
+    <Reference Include="AeroGear.Mobile.Core.Platform.Android">
+      <HintPath>..\..\Core\Core.Platform.Android\bin\Debug\AeroGear.Mobile.Core.Platform.Android.dll</HintPath>
+    </Reference>
+    <Reference Include="AeroGear.Mobile.Core.Platform.Android">
+      <HintPath>..\..\Core\Core.Platform.Android\bin\Debug\AeroGear.Mobile.Core.Platform.Android.dll</HintPath>
+    </Reference>
+    <Reference Include="AeroGear.Mobile.Core.Platform.Android">
+      <HintPath>..\..\Core\Core.Platform.Android\bin\Debug\AeroGear.Mobile.Core.Platform.Android.dll</HintPath>
+    </Reference>
```
Comment (wei-lee) [inline Example/Example.Android/Example.Android.csproj]: Hmm, not sure what's going on here but it looks like these are duplicated refs

File: Example/Example.iOS/Example.iOS.csproj [hunk 7]
```diff
@@ -124,6 +124,57 @@
...
+    </Reference>
+    <Reference Include="AeroGear.Mobile.Core.Platform.iOS">
+      <HintPath>..\..\Core\Core.Platform.iOS\bin\Debug\AeroGear.Mobile.Core.Platform.iOS.dll</HintPath>
+    </Reference>
+    <Reference Include="AeroGear.Mobile.Core.Platform.iOS">
+      <HintPath>..\..\Core\Core.Platform.iOS\bin\Debug\AeroGear.Mobile.Core.Platform.iOS.dll</HintPath>
+    </Reference>
+    <Reference Include="AeroGear.Mobile.Core.Platform.iOS">
+      <HintPath>..\..\Core\Core.Platform.iOS\bin\Debug\AeroGear.Mobile.Core.Platform.iOS.dll</HintPath>
+    </Reference>
```
Comment (wei-lee) [inline Example/Example.iOS/Example.iOS.csproj]: same here.

File: Example/Example.Android/Example.Android.csproj (hunk 6 above)
Comment (aidenkeating) [reply Example/Example.Android/Example.Android.csproj]: Nice catch, not sure how it happened but yep I'll remove them

File: Auth/Auth.Platform.iOS/Credentials/OIDCCredential.cs (hunk 1 above)
Comment (aidenkeating) [reply Auth/Auth.Platform.iOS/Credentials/OIDCCredential.cs]: Yep, I'm guessing @ziccardi copied it across for simpler dev, testing etc.

**Pull request merged**